File: packages/engine/src/core/drive-seen.ts
import { mkdirSync, readFileSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { writeFileAtomic } from './atomic'
import { cloneKey, resolvesInside, stateBase, type StateEnv } from './state-dir'

/**
 * What each session last saw of its initiative's run (drive-visibility 3.2):
 * the mark that lets the prompt's drive line speak only when the run moved.
 *
 * Per-clone state, outside the repo (SPEC §Driver, watching a run): a mark is
 * about one session on one machine, so it is never committed, and a lost one
 * costs a repeated line. The failure direction is the rule here, and it is the
 * opposite of shipwatch's: a mark that cannot be read counts as never seen, and
 * one that cannot be written is news again next prompt, so the line REPEATS —
 * it never goes silent on a run that moved.
 *
 * Sessions sharing a clone share this file. Two prompts in the same instant can
 * revert each other's marks, which again only repeats a line.
 */

/** Bump on ANY change to the on-disk shape — a mismatch reads as no marks. */
export const DRIVE_SEEN_VERSION = 1

/** Marks kept before the least recently marked are evicted; an evicted session sees its line once more. */
export const DRIVE_SEEN_MAX_MARKS = 64

interface Mark {
  seen: string
  seq: number
}

/** `<state base>/drive-seen/<clone key>.json`, or null where that would sit inside the repo. */
export function driveSeenPath(rootDir: string, env: StateEnv = process.env): string | null {
  const path = join(stateBase(env), 'drive-seen', `${cloneKey(rootDir)}.json`)
  return resolvesInside(path, rootDir) ? null : path
}

function readMarks(path: string): Record<string, Mark> {
  let disk: unknown
  try {
    disk = JSON.parse(readFileSync(path, 'utf8'))
  } catch {
    return {}
  }
  if (typeof disk !== 'object' || disk === null) return {}
  const d = disk as { version?: unknown; marks?: unknown }
  if (d.version !== DRIVE_SEEN_VERSION || typeof d.marks !== 'object' || d.marks === null) return {}
  const marks: Record<string, Mark> = {}
  for (const [id, mark] of Object.entries(d.marks as Record<string, unknown>)) {
    const m = mark as Partial<Mark> | null
    if (typeof m?.seen === 'string' && typeof m.seq === 'number' && Number.isFinite(m.seq)) marks[id] = { seen: m.seen, seq: m.seq }
  }
  return marks
}

/**
 * Whether `seen` is news to this session, marking it seen when it is. Writes
 * only on news, so a prompt on an unchanged run costs one small read.
 */
export function noteDriveSeen(rootDir: string, sessionId: string, seen: string, env: StateEnv = process.env): boolean {
  const path = driveSeenPath(rootDir, env)
  if (path === null) return true
  const marks = readMarks(path)
  if (marks[sessionId]?.seen === seen) return false
  let seq = 0
  for (const mark of Object.values(marks)) if (mark.seq > seq) seq = mark.seq
  marks[sessionId] = { seen, seq: seq + 1 }
  const ids = Object.keys(marks)
  const kept: Record<string, Mark> = {}
  for (const id of ids.sort((a, b) => marks[b]!.seq - marks[a]!.seq).slice(0, DRIVE_SEEN_MAX_MARKS)) kept[id] = marks[id]!
  try {
    mkdirSync(dirname(path), { recursive: true })
    writeFileAtomic(path, JSON.stringify({ version: DRIVE_SEEN_VERSION, marks: kept }))
  } catch {
    // Unwritable: this run is news again next prompt, which is the safe direction.
  }
  return true
}

File: crates/sofar-core/src/drive_seen.rs
//! What each session last saw of its initiative's run (`core/drive-seen.ts`,
//! drive-visibility 3.2): the mark that lets the prompt's drive line speak
//! only when the run moved. Per-clone state outside the repo. A mark that
//! cannot be read counts as never seen, and one that cannot be written is
//! news again next prompt: the line repeats, it never goes silent.

use std::path::{Path, PathBuf};

use crate::atomic::write_file_atomic;
use crate::diagnostics::{clone_key, resolves_inside, state_base};
use crate::json::{self, Json, Object};

/// `DRIVE_SEEN_VERSION` / `DRIVE_SEEN_MAX_MARKS`.
pub const DRIVE_SEEN_VERSION: f64 = 1.0;
pub const DRIVE_SEEN_MAX_MARKS: usize = 64;

/// `driveSeenPath`: `<state base>/drive-seen/<clone key>.json`, or None
/// where that would sit inside the repo.
#[must_use]
pub fn drive_seen_path(root: &Path) -> Option<PathBuf> {
    let path = state_base()
        .join("drive-seen")
        .join(format!("{}.json", clone_key(root)));
    (!resolves_inside(&path, root)).then_some(path)
}

/// `readMarks`: session → (seen, seq), in the file's (JavaScript) order.
fn read_marks(path: &Path) -> Vec<(String, String, f64)> {
    let Ok(bytes) = std::fs::read(path) else {
        return Vec::new();
    };
    let Ok(Json::Obj(disk)) = json::parse(&String::from_utf8_lossy(&bytes)) else {
        return Vec::new();
    };
    if disk.get("version") != Some(&Json::Num(DRIVE_SEEN_VERSION)) {
        return Vec::new();
    }
    let Some(Json::Obj(marks)) = disk.get("marks") else {
        return Vec::new();
    };
    marks
        .js_ordered()
        .into_iter()
        .filter_map(|(id, mark)| {
            let m = mark.as_obj()?;
            let seen = m.get("seen")?.as_str()?;
            let seq = m.get("seq")?.as_f64().filter(|n| n.is_finite())?;
            Some((id.to_owned(), seen.to_owned(), seq))
        })
        .collect()
}

/// `noteDriveSeen`: whether `seen` is news to this session, marking it seen
/// when it is. Writes only on news.
pub fn note_drive_seen(root: &Path, session_id: &str, seen: &str) -> bool {
    let Some(path) = drive_seen_path(root) else {
        return true;
    };
    let mut marks = read_marks(&path);
    if marks.iter().any(|(id, s, _)| id == session_id && s == seen) {
        return false;
    }
    let seq = marks.iter().map(|(_, _, q)| *q).fold(0.0_f64, f64::max);
    match marks.iter_mut().find(|(id, _, _)| id == session_id) {
        Some(mark) => {
            seen.clone_into(&mut mark.1);
            mark.2 = seq + 1.0;
        }
        None => marks.push((session_id.to_owned(), seen.to_owned(), seq + 1.0)),
    }
    // Most recently marked first (stable), the least recent evicted.
    marks.sort_by(|a, b| b.2.total_cmp(&a.2));
    marks.truncate(DRIVE_SEEN_MAX_MARKS);
    let mut kept = Object::with_capacity(marks.len());
    for (id, seen, seq) in marks {
        let mut m = Object::with_capacity(2);
        m.insert("seen", Json::Str(seen));
        m.insert("seq", Json::Num(seq));
        kept.insert(id, Json::Obj(m));
    }
    let mut o = Object::with_capacity(2);
    o.insert("version", Json::Num(DRIVE_SEEN_VERSION));
    o.insert("marks", Json::Obj(kept));
    if let Some(dir) = path.parent() {
        let _ = std::fs::create_dir_all(dir);
    }
    let _ = write_file_atomic(&path, json::stringify(&Json::Obj(o)).as_bytes());
    true
}
